### PR TITLE
Authentication Service

### LINKED ISSUE #5: Authentication service (12H)
Requires: #8 .

### PR DESCRIPTION
**Pull request by calesanz**
Implemented login and registration.

Closes #5

### CODE REVIEW

**Comment by calesanz** [comment]
The signup and login could be used for the demo.

File: source/README.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ Grants should be given to the user `social_anxiety` identified by `social_anxiet
 
 If you're using Docker it will be a one liner:
 ```bash
-docker run --name=[GIVE_IT_A_NAME] -e POSTGRES_USER=social_anxiety -e POSTGRES_PASSWORD=social_anxiety_password -p 5432:5432 postgresql
+a
```
Comment (calesanz) [inline source/README.md:15]: My bad, i have to revert this change.

File: source/src/main/resources/application.yml [hunk 2]
```diff
@@ -5,10 +5,10 @@ server:
 # Database configuration
 spring:
   datasource:
-    url: ${APPl_DATASOURCE_JDBC_URL}
```
Comment (calesanz) [inline source/src/main/resources/application.yml:8]: I have to undo the changes to this file too.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/Patient.java [hunk 3]
```diff
@@ -20,12 +23,13 @@
 public class Patient {
 
   @Id
+  @GeneratedValue(strategy= GenerationType.IDENTITY)
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/Patient.java:26]: This must be `GenerationType.SEQUENCE` in order to let PostgreSQL generate a value. I've an example [here](https://github.com/tbmelabs/tbmelabs-authorization-server/blob/development/src/main/java/ch/tbmelabs/authorizationserver/domain/User.java). It will be:
```java
public class Patient {

  [...]

  @Id
  @GenericGenerator(name = "pk_sequence",
    strategy = AbstractAuditingEntity.SEQUENCE_GENERATOR_STRATEGY,
    parameters = {@Parameter(name = "sequence_name", value = "patient_id_seq"),
      @Parameter(name = "increment_size", value = "1")})
  @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "pk_sequence")
  private Long id;

  [...]
```
the `patient_id_seq` was generated because of the PostgreSQL type `BIGSERIAL` (an "autoincrementing eight-byte integer"). documentation is [here](https://www.postgresql.org/docs/12/datatype.html).

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/PatientRepository.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package ch.bfh.bti7081.s2020.blue.domain;
+
+import org.springframework.data.repository.CrudRepository;
+
+public interface PatientRepository extends CrudRepository<Patient,String> {
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/PatientRepository.java:5]: naming maybe.. and you didn't format this file: `CrudRepository<Patient,_String>`

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterDto.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
+package ch.bfh.bti7081.s2020.blue.domain;
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterDto.java:1]: it's not a domain class. maybe `view.dto`, but `domain` is reserved for (persistent) entities.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterDto.java [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  private String password;
+  @Size(min = 6, max = 128)
+  private String repeatPassword;
+  @Email
+  private String email;
+  @Size(min = 2, max = 128)
+  private String surname;
+  @Size(min = 2, max = 128)
+  private String givenName;
+
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterDto.java:27]: use a blank line between all attributes?

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+@Service
+@Log4j2
+public class RegisterService {
+
+  private final LoginRepository loginRepository;
+  private final PatientRepository patientRepository;
+
+  public RegisterService(
+      @Autowired LoginRepository loginRepository,
+      @Autowired PatientRepository patientRepository) {
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java:20]: `@Autowired` is implicit because of the `@Service` annotation. you cannot have constructor arguments which are not autowired - please remove it.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
+package ch.bfh.bti7081.s2020.blue.domain;
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java:1]: it's not a domain class. use `ch.bfh.bti7081.s2020.blue.service`.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+  public RegisterService(
+      @Autowired LoginRepository loginRepository,
+      @Autowired PatientRepository patientRepository) {
+    this.loginRepository = loginRepository;
+    this.patientRepository = patientRepository;
+
+  }
+
+  public Collection<ValidationError> register(RegisterDto registerDto) {
+    var violations = Validation.buildDefaultValidatorFactory().getValidator().validate(registerDto);
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java:27]: please do not use `var`. we're in an object oriented world - not js. also, if chaining commands, always add a newline (like you did with the stream below).

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    var validationErrors = violations.stream()
+        .map(violation -> new ValidationError(
+            String.format("%s: %s", violation.getPropertyPath(), violation.getMessage())))
+        .collect(Collectors.toList());
+
+    try {
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java:33]: this try block is too long. can you move it into a separate method?

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+      if (!isUsernameUnique(registerDto.getUsername())) {
+        validationErrors.add(new ValidationError("Username is already in use."));
+      }
+      if (!isEmailUnique(registerDto.getEmail())) {
+        validationErrors.add(new ValidationError("E-Mail address is already in use."));
+      }
+      if (!registerDto.getPassword().equals(registerDto.getRepeatPassword())) {
+        validationErrors.add(new ValidationError("Passwords did not match."));
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java:41]: "Passwords *do* not match."
when filling in the form, the user is in "the now". neither the past nor the future ;)

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+      }
+
+      if (validationErrors.isEmpty()) {
+        var password = new BCryptPasswordEncoder().encode(registerDto.getPassword());
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java:45]: do not initialize `BCryptPasswordEncoder` for every registration. waist of memory..

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginView.java [hunk 13]
```diff
@@ -0,0 +1,10 @@
+package ch.bfh.bti7081.s2020.blue.view.authentication;
+
+public interface LoginView {
+
+  void addListener(LoginViewListener listener);
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginView.java:5]: remove.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginViewImpl.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+@Route(value = LoginViewImpl.ROUTE)
+@PageTitle("Login")
+public class LoginViewImpl extends VerticalLayout implements BeforeEnterObserver, LoginView {
+
+  static final String ROUTE = "login";
+  private List<LoginViewListener> listeners = new ArrayList<>();
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginViewImpl.java:19]: this can be a single listener.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginViewImpl.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  private LoginForm login = new LoginForm();
+
+  public LoginViewImpl() {
+    new LoginPresenter(null, this);
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginViewImpl.java:23]: directly assign to `this.listener`.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/presenter/LoginPresenter.java [hunk 16]
```diff
@@ -0,0 +1,11 @@
+package ch.bfh.bti7081.s2020.blue.presenter;
+
+import ch.bfh.bti7081.s2020.blue.view.authentication.LoginView;
+import ch.bfh.bti7081.s2020.blue.view.authentication.LoginView.LoginViewListener;
+
+public class LoginPresenter implements LoginViewListener {
+
+  public LoginPresenter(Object model, LoginView view) {
+    view.addListener(this);
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/presenter/LoginPresenter.java:9]: the view should not have `addListener`.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/presenter/RegisterPresenter.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    view.showMessage("");
+    var errors = registerService.register(model);
+    if (errors.isEmpty()) {
+      view.navigate("");
+    } else {
+      view.showMessage(
+          errors.stream()
+              .map(ValidationError::getMessage)
+              .collect(Collectors.joining(" ")));
+
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/presenter/RegisterPresenter.java:35]: trailing newline. (remove)

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/security/SecurityConfiguration.java [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+@Configuration
+public class SecurityConfiguration extends WebSecurityConfigurerAdapter {
+
+  private static final String LOGIN_PROCESSING_URL = "/login";
+  private static final String LOGIN_FAILURE_URL = "/login?error";
+  private static final String LOGIN_URL = "/login";
+  private static final String LOGOUT_SUCCESS_URL = "/login";
+  private static final String LOGOUT_URL = "/logout";
+  @Autowired
+  private UserDetailsService databaseUserDetailsService;
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/security/SecurityConfiguration.java:24]: use constructor injection. it's much more testable.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/security/SecurityConfiguration.java [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Require login to access internal pages and configure login form.
+   */
+  @Override
+  protected void configure(HttpSecurity http) throws Exception {
+    http.anonymous().and().authorizeRequests()
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/security/SecurityConfiguration.java:31]: call to `anonymous()` is unnecessary.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/security/DatabaseUserDetailsService.java [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+
+@Service
+public class DatabaseUserDetailsService implements UserDetailsService {
+  @Autowired
+  private LoginRepository loginRepository;
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/security/DatabaseUserDetailsService.java:14]: use constructor injection.

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/security/CustomRequestCache.java [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+
+import org.springframework.security.web.savedrequest.HttpSessionRequestCache;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * HttpSessionRequestCache that avoids saving internal framework requests.
+ */
+class CustomRequestCache extends HttpSessionRequestCache {
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/security/CustomRequestCache.java:11]: this name is not really self-describing.
and is our session storage in memory?

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/security/SecurityUtils.java [hunk 22]
```diff
@@ -0,0 +1,47 @@
+package ch.bfh.bti7081.s2020.blue.security;
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/security/SecurityUtils.java:1]: so, shouldn't this be a `util` (-package) class?

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginViewImpl.java [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+    login.setAction("login");
+    this.add(login);
+  }
+
+  @Override
+  public void beforeEnter(BeforeEnterEvent event) { //
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/LoginViewImpl.java:29]: use `HasUrlParameter`: [example from @Marcarrian](https://github.com/zeroplexer/ch.bfh.bti7081.s2020.blue/blob/feature/%2321-grundstruktur-komponenten/source/src/main/java/ch/bfh/bti7081/s2020/blue/view/journal/JournalDetailViewImpl.java).

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/RegisterViewImpl.java [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+    var emailField = new EmailField();
+    emailField.setValueChangeMode(ValueChangeMode.EAGER);
+    addFormItem(emailField, "E-Mail");
+    binder.forField(emailField).withValidator(new EmailValidator(
+        "This doesn't look like a valid email address"))
+        .withValidator(listener::isEmailUnique, "Email address is already registered.")
+        .bind(RegisterDto::getEmail, RegisterDto::setEmail);
+
+
+
```
Comment (bbortt) [inline source/src/main/java/ch/bfh/bti7081/s2020/blue/view/authentication/RegisterViewImpl.java:78]: trailing newlines. also from lines 86 - 87.
do I need to say I hate Vaadin? how should a view like this ever get readable? 😡

File: source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java (hunk 9 above)
Comment (Marcarrian) [reply source/src/main/java/ch/bfh/bti7081/s2020/blue/domain/RegisterService.java:27]: Würde ich nicht einfach so generell sagen. `var` kann man dort verwenden, wo die typen überhaupt keine Rolle spielen. 'Overusen' finde ich natürlich auch nicht gut. Aber das ist eine Philosophie für sich und da finde ich es einfach sehr unkonstruktiv zu sagen verwende **nie** `var`.
